### PR TITLE
Week-1 - Implementation of ErrorOr

### PR DESCRIPTION
**Pull request by GeorgiyKozak**

### CODE REVIEW

File: src/main/scala/karazin/scala/users/group/week1/homework/adt.scala [hunk 1]
```diff
@@ -10,26 +10,31 @@ package karazin.scala.users.group.week1.homework
...
       
       Make sure that in case of failing the method with exception
       no exception is thrown but the case for an error is returned
-    */ 
-    def flatMap = ???
+    */
+    def flatMap[Q](f: V => ErrorOr[Q]): ErrorOr[Q] =
+      this match
+        case ErrorOr.Error(ex) => ErrorOr.Error(ex)
+        case ErrorOr.Some(v)   => try f(v) catch {
```
Comment (IgorWolkov) [inline src/main/scala/karazin/scala/users/group/week1/homework/adt.scala:34]: Nobody did it with `try-catch` at the first attempt. I hope you got the idea why `f(v)` should be wrapped into `try-catch`.

File: src/main/scala/karazin/scala/users/group/week1/homework/adt.scala [hunk 2]
```diff
@@ -48,6 +58,7 @@ object adt:
       Make sure that in case of failing the method with exception
       no exception is thrown but the case for an error is returned
     */
-    def apply = ???
+    def apply[V](v: V): ErrorOr[V] =
+      if v == null then ErrorOr.Error(NullPointerException()) else ErrorOr.Some(v)
```
Comment (IgorWolkov) [inline src/main/scala/karazin/scala/users/group/week1/homework/adt.scala]: Always provide a useful message when throwing/passing an exception.

**Review by IgorWolkov — CHANGES_REQUESTED** [review]
Minor comment.

File: src/main/scala/karazin/scala/users/group/week1/homework/adt.scala (hunk 1 above)
Comment (GeorgiyKozak) [reply src/main/scala/karazin/scala/users/group/week1/homework/adt.scala:34]: In the last lecture you said that we should consider that the method f(v) can throw an exception. f(v) is some function in the body of which an exception can be thrown, right?

Comment (GeorgiyKozak) [reply src/main/scala/karazin/scala/users/group/week1/homework/adt.scala:34]: In the last lecture you said that we should consider that the function f(v) can throw an exception. f(v) is some function in the body of which an exception can be thrown, right?

Comment (GeorgiyKozak) [reply src/main/scala/karazin/scala/users/group/week1/homework/adt.scala:34]: In the last lecture you said that we should consider that the function f(v) can throw an exception. f(v) is some function in the body of which an exception can be thrown or f(v) can return ErrorOr.Error?

Comment (GeorgiyKozak) [reply src/main/scala/karazin/scala/users/group/week1/homework/adt.scala:34]: In the last lecture you said that we should consider that the function f(v) can throw an exception. f(v) is some function in the body of which an exception can be thrown or f(v) can return ErrorOr.Error?

**Review by IgorWolkov — APPROVED** [review]
Well done!

**Pull request merged**